### PR TITLE
Fix AMGReduced

### PR DESCRIPTION
**Pull request by mschreter**
... only a "minifix". I'll merge once CI is happy. @nmuch FYI: `AMGReduced` should now work without the exception. ~~However, I got into troubles when using `AMG` or `AMGReduced` together with AMR. I obtain the following stacktrace:~~

```bash
Stacktrace:
-----------
#0  /home/magdalena/TUM/build-deal/lib/libdeal_II.g.so.9.3.0-pre: dealii::TrilinosWrappers::SparseMatrix::add(unsigned int, unsigned int, unsigned int const*, double const*, bool, bool)
#1  /home/magdalena/TUM/build-deal/lib/libdeal_II.g.so.9.3.0-pre: void dealii::AffineConstraints<double>::distribute_local_to_global<dealii::TrilinosWrappers::SparseMatrix, dealii::Vector<double> >(dealii::FullMatrix<double> const&, dealii::Vector<double> const&, std::vector<unsigned int, std::allocator<unsigned int> > const&, dealii::TrilinosWrappers::SparseMatrix&, dealii::Vector<double>&, bool, std::integral_constant<bool, false>) const
#2  meltpooldg: void dealii::AffineConstraints<double>::distribute_local_to_global<dealii::TrilinosWrappers::SparseMatrix>(dealii::FullMatrix<double> const&, std::vector<unsigned int, std::allocator<unsigned int> > const&, dealii::TrilinosWrappers::SparseMatrix&) const
#3  meltpooldg: MeltPoolDG::Heat::HeatTransferOperator<2, double>::compute_system_matrix(dealii::TrilinosWrappers::SparseMatrix&, bool) const
#4  meltpooldg: MeltPoolDG::Heat::HeatTransferOperation<2>::solve(double)::{lambda(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&)#2}::operator()(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&) const
#5  meltpooldg: int std::__invoke_impl<int, MeltPoolDG::Heat::HeatTransferOperation<2>::solve(double)::{lambda(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&)#2}&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&>(std::__invoke_other, MeltPoolDG::Heat::HeatTransferOperation<2>::solve(double)::{lambda(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&)#2}&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&)
#6  meltpooldg: std::enable_if<is_invocable_r_v<int, MeltPoolDG::Heat::HeatTransferOperation<2>::solve(double)::{lambda(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&)#2}&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&>, std::enable_if>::type std::__invoke_r<int, MeltPoolDG::Heat::HeatTransferOperation<2>::solve(double)::{lambda(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&)#2}&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&>(int&&, (MeltPoolDG::Heat::HeatTransferOperation<2>::solve(double)::{lambda(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii
--------------------------------------------------------
An error occurred in line <1692> of file </home/magdalena/TUM/dealii/source/lac/trilinos_sparse_matrix.cc> in function
    void dealii::TrilinosWrappers::SparseMatrix::add(dealii::TrilinosWrappers::SparseMatrix::size_type, dealii::TrilinosWrappers::SparseMatrix::size_type, const size_type*, const TrilinosScalar*, bool, bool)
The violated condition was: 
    nonlocal_matrix->RowMap().LID( static_cast<TrilinosWrappers::types::int_type>(row)) != -1
Additional information: 
    Attempted to write into off-processor matrix row that has not be
    specified as being writable upon initialization
```

~~Dear @peterrum: Do you have any suspisions on that?~~

### CODE REVIEW

File: include/meltpooldg/heat/heat_transfer_operation.hpp [hunk 1]
```diff
@@ -166,7 +166,7 @@ namespace MeltPoolDG::Heat
 
             auto preconditioner =
               LinearSolve::setup_preconditioner(heat_transfer_preconditioner.get_system_matrix(),
-                                                heat_data.solver.preconditioner_type);
+                                                "AMG");
```
Comment (peterrum) [inline include/meltpooldg/heat/heat_transfer_operation.hpp:169]: At some time we might want to enable ILU and Identity again.

**Comment by mschreter** [comment]
sorry, I've messed up the commit history. I'll fix this issue soon.

**Comment by peterrum** [comment]
> Do you have any suspisions on that?

It's hard to say. It looks like some "locally-relevant entries" are missing from the matrix. Could you disable to following lines:

https://github.com/MeltPoolDG/MeltPoolDG/blob/f765c1e81927dd9dd4e4f568d4dd6c13e55459cb/include/meltpooldg/heat/heat_transfer_preconditioner.hpp#L52-L56

If that does not work, I can take a look later.

File: include/meltpooldg/heat/heat_transfer_operation.hpp (hunk 1 above)
Comment (mschreter) [reply include/meltpooldg/heat/heat_transfer_operation.hpp]: I've re-enabled ILU and introduced ILUReduced. Identity has still existed.

File: include/meltpooldg/heat/heat_transfer_operation.hpp [hunk 2]
```diff
@@ -159,14 +159,18 @@ namespace MeltPoolDG::Heat
...
             heat_operator->compute_system_matrix(heat_transfer_preconditioner.get_system_matrix(),
-                                                 heat_data.solver.preconditioner_type == "AMG");
+                                                 heat_data.solver.preconditioner_type == "AMG" ||
+                                                   heat_data.solver.preconditioner_type == "ILU");
 
+            // take the first three letters as relevant preconditioner type
             auto preconditioner =
               LinearSolve::setup_preconditioner(heat_transfer_preconditioner.get_system_matrix(),
-                                                "AMG");
+                                                heat_data.solver.preconditioner_type.substr(0, 3));
```
Comment (peterrum) [inline include/meltpooldg/heat/heat_transfer_operation.hpp]: This is ugly ;)

Comment (mschreter) [reply include/meltpooldg/heat/heat_transfer_operation.hpp]: Haha, feel free to point out a 

> > Do you have any suspisions on that?
> 
> It's hard to say. It looks like some "locally-relevant entries" are missing from the matrix. Could you disable to following lines:
> 
> https://github.com/MeltPoolDG/MeltPoolDG/blob/f765c1e81927dd9dd4e4f568d4dd6c13e55459cb/include/meltpooldg/heat/heat_transfer_preconditioner.hpp#L52-L56
> 
> If that does not work, I can take a look later.

Thanks, I've fixed it (was one of my corpses in the cellar ;-)).

Comment (peterrum) [reply include/meltpooldg/heat/heat_transfer_operation.hpp]: Personally, I would prefer `heat_data.solver.preconditioner_type=="AMGReduaced" : heat_data.solver.preconditioner_type`.

> Thanks, I've fixed it (was one of my corpses in the cellar ;-)).

This is an more obvious reason^^ I would have been very surprised if that function did not work.

Comment (mschreter) [reply include/meltpooldg/heat/heat_transfer_operation.hpp]: Mmh. I saw your comment too late. Do you also agree with my current solution :-)?

**Review by peterrum — APPROVED** [review]
OK. I'll merge later!

**Pull request merged**